corehq/pillows/case_search.py

```diff
@@ -0,0 +1,151 @@
+        {'key': 'external_id', 'value': doc_dict.get('external_id')}
+    ]
+    dynamic_case_properties = [
+        {'key': key, 'value': value}
+        for key, value in doc_dict.iteritems()
+        if _is_dynamic_case_property(key)
+    ]
+
+    return base_case_properties + dynamic_case_properties
+
+
+def _is_dynamic_case_property(prop):
+    """
+    Finds whether {prop} is a dynamic property of CommCareCase. If so, it is likely a case property.
+    """
+    return not inspect.isdatadescriptor(getattr(CommCareCase, prop, None)) and re.search(r'^[a-zA-Z]', prop)
+
+
+class CaseSearchPillowProcessor(ElasticProcessor):
+    def process_change(self, pillow_instance, change, do_set_checkpoint):
+        domain = self._get_domain(change)
+        change_object = self._get_change_object(change)
+        if domain and case_search_enabled_for_domain(domain):
+            super(CaseSearchPillowProcessor, self).process_change(
+                pillow_instance, change_object, do_set_checkpoint
+            )
+
+    def _get_domain(self, change):
+        if isinstance(change, dict):
+            return change.get('key', [])[0]
+        elif isinstance(change, Change):
+            return change.metadata.domain
+
+    def _get_change_object(self, change):
+        if isinstance(change, dict):
+            return change_from_couch_row(change)
+        elif isinstance(change, Change):
+            return change
+
+
+def get_case_search_to_elasticsearch_pillow(pillow_id='CaseSearchToElasticsearchPillow'):
+    checkpoint = PillowCheckpoint(
+        'case-search-to-elasticsearch',
+    )
+    case_processor = CaseSearchPillowProcessor(
+        elasticseach=get_es_new(),
+        index_meta=ElasticsearchIndexMeta(index=CASE_SEARCH_INDEX, type=CASE_ES_TYPE),
+        doc_prep_fn=transform_case_for_elasticsearch
+    )
+    return ConstructedPillow(
+        name=pillow_id,
+        document_store=None,
+        checkpoint=checkpoint,
+        change_feed=KafkaChangeFeed(topics=[topics.CASE], group_id='cases-to-es'),
+        processor=case_processor,
+        change_processed_event_handler=PillowCheckpointEventHandler(
+            checkpoint=checkpoint, checkpoint_frequency=100,
+        ),
+    )
+
+
+def get_couch_case_search_reindexer(domain=None):
+    # TODO: Figure out how to not fetch every single case from the DB when running a full reindex
+
+    # TODO: Remove this
+    # It initializes the es index if it doesn't exist
+    CaseSearchPillow()
+
+    view_kwargs = {'include_docs': True}
+
+    if domain is not None:
+        if not case_search_enabled_for_domain(domain):
+            raise CaseSearchNotEnabledException("{} does not have case search enabled".format(domain))
+
+        startkey = [domain]
+        endkey = [domain, {}, {}]
+        view_kwargs.update({'startkey': startkey, 'endkey': endkey})
+
+    return PillowReindexer(get_case_search_to_elasticsearch_pillow(), CouchViewChangeProvider(
```

Comment:
I don't think this abstraction layer exists yet... I imagine it would be a ChangeProvider that would yield both Couch changes and SQL changes (i.e. smooshing together `CouchViewChangeProvider` and `SqlCaseChangeProvider) so a full reindex would provide all cases. 

It would be great to talk about the best way of going about this in general though. I couldn't figure out a way to pull from couch all the cases that belong to a _list_ of domains (i.e. something like `WHERE domain IN (all_case_search_enabled_domains)`. With the SQL dbaccessors we have, I'm also not sure how we would do that either. (It might not be possible with couch, and might not be worth the effort)

In the instance where the domain is passed in, I think I'd have to make a new type of `SQLCaseChangeProvider` which can take in a dbaccessor method to filter down the cases it yields.